diff -ur a/packages/services/Telecomm/src/com/android/server/telecom/CallAudioManager.java b/packages/services/Telecomm/src/com/android/server/telecom/CallAudioManager.java
--- a/packages/services/Telecomm/src/com/android/server/telecom/CallAudioManager.java	Wed Apr 20 08:50:20 2016
+++ b/packages/services/Telecomm/src/com/android/server/telecom/CallAudioManager.java	Thu May 12 01:31:40 2016
@@ -134,9 +134,20 @@
                     Call call = mCallsManager.getForegroundCall();
                     boolean setMsimAudioParams = SystemProperties
                             .getBoolean("ro.multisim.set_audio_params", false);
+					boolean setSamsungRealcallParams = SystemProperties
+							.getBoolean("ro.telephony.samsung.realcall", false);		
 
                     Log.v(this, "Request to change audio mode from %s to %s", modeToString(oldMode),
                             modeToString(newMode));
+							
+					if(setSamsungRealcallParams) {
+						if(newMode == 0) {
+							mAudioManager.setParameters("realcall=off");
+						} else if(newMode == AudioManager.MODE_IN_COMMUNICATION || newMode == AudioManager.MODE_RINGTONE
+						|| newMode == AudioManager.MODE_IN_CALL) {
+							mAudioManager.setParameters("realcall=on");
+						}
+					}		
 
                     if (oldMode != newMode) {
                         if (oldMode == AudioManager.MODE_IN_CALL &&
